File: Medium/Maximum Swap.py
# https://leetcode.com/problems/maximum-swap
# TC: O(N)
# SC: O(1)
class Solution:
    def maximumSwap(self, num: int) -> int:
        if num < 10: return num

        num = list(str(num))

        # Store the last index of each digit
        last = [None] * 10
        for i, d in enumerate(num):
            d = ord(d) - 48
            last[d] = i

        # Find the best swap, if any
        for i, d in enumerate(num):
            d = ord(d) - 48

            # Check for the largest number larger than d and that comes after d, if any
            for j in range(9, d, -1):
                if last[j] and last[j] > i:
                    # If we found a greater number, swap
                    num[i], num[last[j]] = num[last[j]], num[i]
                    return int(''.join(num))
        
        # if no swaps could be made (in the case of a non-increasing set of digits I think)
        # just return the number
        return int(''.join(num))

File: TypeScript/Medium/Maximum Swap.ts
// https://leetcode.com/problems/maximum-swap

function maximumSwap(num: number): number {
    if (num < 10) return num;

    let num_str: any = '' + num;
    let dig, idx;

    // Find the last index of each number from 0 - 9
    let last: Array<number> = Array(10);
    for (let idx in num_str) {
        dig = +num_str[idx]; // + converts to number
        last[dig] = +idx;
    }

    // Look for the best swap that can be made, if any
    for (let idx_str in num_str) {
        idx = +idx_str;  // convert to number
        dig = +num_str[idx]

        // Search for largest number that is at a later position, if any
        for (let j = 9; j > dig; j--) {
            if (last[j] && last[j] > +idx) {
                // Found the max swap. Swap and return.
                return  +(num_str.substring(0, idx) 
                        + (''+num_str[last[j]]) 
                        + num_str.substring(idx+1, last[j]) 
                        + (''+num_str[idx]) 
                        + num_str.substring(last[j] + 1))
            }
        }
    }

    // if no swaps could be made, just return the number
    return +num;
};
